pennylane/utils.py

```diff
@@ -105,6 +106,56 @@ def decompose_hamiltonian(H, hide_identity=False):
     return coeffs, obs
 
 
+def sparse_hamiltonian(H):
+    r"""Computes the sparse matrix representation a Hamiltonian in the computational basis.
+
+    Args:
+        H (~.Hamiltonian): Hamiltonian operator for which the matrix representation should be
+         measured
+
+    Returns:
+        coo_matrix: a sparse matrix in scipy COOrdinate format with the dimension of
+        :math:`(2^n, 2^n)` where :math:`n` is the number of wires
+
+    **Example:**
+
+    This function can be used by passing a `qml.Hamiltonian` object as:
+
+    >>> coeffs = [0.5, 0.5]
+    >>> obs = [qml.PauliZ(wires=[0]) @ qml.PauliZ(wires=[1]),
+    ...        qml.Identity(wires=[0]) @ qml.PauliZ(wires=[1])]
+    >>> H = qml.Hamiltonian(coeffs, obs)
+    >>> H_sparse = sparse_hamiltonian(H)
+
+    The resulting sparse matrix can be either used directly or transformed into a numpy array:
+
+    >>> H_sparse.toarray()
+    array([[ 1.+0.j,  0.+0.j,  0.+0.j,  0.+0.j],
+           [ 0.+0.j, -1.+0.j,  0.+0.j,  0.+0.j],
+           [ 0.+0.j,  0.+0.j,  0.+0.j,  0.+0.j],
+           [ 0.+0.j,  0.+0.j,  0.+0.j,  0.+0.j]])
+    """
+    if not isinstance(H, qml.Hamiltonian):
+        raise TypeError("Passed Hamiltonian must be of type `qml.Hamiltonian`")
+
+    n = len(H.wires)
+
+    matrix = scipy.sparse.coo_matrix((2 ** n, 2 ** n), dtype="complex128")
+
+    for coeffs, ops in zip(H.coeffs, H.ops):
+
+        obs = [scipy.sparse.coo_matrix(o.matrix) for o in ops.obs]
+
+        mat = [scipy.sparse.eye(2, format="coo")] * n
+
+        for i, j in enumerate(ops.wires):
+            mat[j] = obs[i]
+
+        matrix += functools.reduce(lambda i, j: scipy.sparse.kron(i, j, format="coo"), mat) * coeffs
+
+    return matrix.tocoo()
```

Comment:
Is `matrix` not already in COO format? all the elements on `mat` are sparse in COO format, so wouldn't their tensor product also be in this format? :thinking:

Reply:
Scipy changes the format for some arithmetic such as matrix addition (not sure why). I made that conversion in the return to make sure such changes do not affect the final format.